File: ExecutionContext/MultipleOrderedEC-Python/MultipleOrderedEC_Lib/MPTask.py
# -*- coding: utf-8 -*-

##
#   @file MPTask.py
#   @brief GUI、直列ブロック実行スレッド

import threading
import time

import OpenRTM_aist


pyqtExist = True
try:
    from PyQt4 import QtCore, QtGui
except:
    pyqtExist = False


if pyqtExist:
    from MainWindow import MainWindow



##
# @class MPComp
# @brief 同一スレッドでの実行順序クラス
#
class MPComp:
    ##
    # @brief コンストラクタ
    # @param self 
    def __init__(self):
        self.comp = None
        self.I = 0
        self.J = 0
        self.K = 0


##
# @class MPTask
# @brief 直列ブロックを実行するスレッドのクラス
#
class MPTask(OpenRTM_aist.Task):
    ##
    # @brief コンストラクタ
    # @param self
    # @param c 実行コンテキストオブジェクト
    def __init__(self, c):
        OpenRTM_aist.Task.__init__(self)
        self.m_ec = c
        self.m_comp = []
    
    ##
    # @brief コンポーネントを追加する関数
    # @param self
    # @param c ブロック
    # @param I　全体の並列ブロックの番号
    # @param J 並列ブロック内の直列ブロックの番号
    # @param K 直列ブロック内のブロックの番号
    def addComp(self, c, I, J, K):
        self.mc = MPComp()
        self.mc.comp = c
        self.mc.I = I
        self.mc.J = J
        self.mc.K = K
        self.m_comp.append(self.mc)
    
    ##
    # @brief スレッド実行関数
    # @param self
    # @return 0
    def svc(self):
        if len(self.m_ec.rs) > self.m_ec.r_num:
            for i in range(0, len(self.m_comp)):
                self.m_ec.rs[self.m_ec.r_num].rs[self.m_comp[i].I].SR[self.m_comp[i].J][self.m_comp[i].K].s = 1
                self.m_ec.workerComp(self.m_comp[i].comp)
                self.m_ec.rs[self.m_ec.r_num].rs[self.m_comp[i].I].SR[self.m_comp[i].J][self.m_comp[i].K].s = 0

        return 0



##
# @class GUITask
# @brief GUIを実行するスレッド
#
class GUITask(OpenRTM_aist.Task):
    app_flag = False
    ##
    # @brief コンストラクタ
    # @param self 
    # @param ec 実行コンテキストオブジェクト
    def __init__(self, ec):
        OpenRTM_aist.Task.__init__(self)
        self.m_ec = ec
    ##
    # @brief スレッド実行関数
    # @param self
    # @return 0
    def svc(self):
        if GUITask.app_flag == False:
            GUITask.app_flag = True
            guard = OpenRTM_aist.ScopedLock(self.m_ec._workerthread._mutex)

            app = QtGui.QApplication([""])
            mainWin = MainWindow(self.m_ec)
            mainWin.show()
            
            del guard

            app.exec_()
        else:
            pass

        return 0

    ##
    # @brief 未使用
    # @param self
    def updateRTC(self):
        pass

File: ExecutionContext/MultipleOrderedEC-C++/MPTask.cpp
/*!
* @file  MPTask.cpp
* @brief GUI、直列ブロック実行スレッド
*
*/


#include "MPTask.h"

bool GUITask::app_flag = false;

/**
*@brief 直列ブロックを実行するスレッドのクラスのコンストラクタ
*/
MPTask::MPTask(RTC::MultipleOrderedEC *c)
{
	m_ec = c;
}

/**
*@brief
* @param c ブロック
* @param I 全体の並列ブロックの番号
* @param J 並列ブロック内の直列ブロックの番号
* @param K 直列ブロック内のブロックの番号
*/
void MPTask::addComp(sub_Rule *c, int I, int J, int K)
{
	MPComp mc;
	mc.comp = c;
	mc.I = I;
	mc.J = J;
	mc.K = K;
	m_comp.push_back(mc);
	
}

/**
*@brief スレッド実行関数
* @return
*/
int MPTask::svc()
{
	if(m_ec->rs.size() > m_ec->r_num)
	{
		for(int i=0;i < m_comp.size();i++)
		{
			m_ec->rs[m_ec->r_num].rs[m_comp[i].I].SR[m_comp[i].J][m_comp[i].K].s = 1;
			m_ec->workerComp(m_comp[i].comp);
			m_ec->rs[m_ec->r_num].rs[m_comp[i].I].SR[m_comp[i].J][m_comp[i].K].s = 0;
			
		}
	}

	return 0;
}

/**
*@brief GUIを実行するスレッドのコンストラクタ
* @param ec 実行コンテキストオブジェクト
*/
GUITask::GUITask(RTC::MultipleOrderedEC *ec)
{
	
	this->m_ec = ec;
	
	
}

/**
*@brief スレッド実行関数
* @return
*/
int GUITask::svc()
{
	
#ifdef QT_GUI_LIB
	if(!app_flag)
	{
		app_flag = true;

		mu.lock();

		int _argc(1);
		char* _argv[]={""};

		Q_INIT_RESOURCE(MultipleOrderedEC);
		app = new QApplication(_argc, (char**)_argv);

		

		mw = new MainWindow(this->m_ec);

		

		mu.unlock();

		mw->show();

		

		app->exec();
	}
	else
	{
		
	}
#endif
	return 0;

	
}

/**
*@brief 未使用
* @return
*/
void GUITask::updateRTC()
{

}
